batch/sql/estimated-current.sql

```diff
@@ -187,6 +188,40 @@ CREATE INDEX `batches_token` ON `batches` (`token`);
 CREATE INDEX `batches_time_completed` ON `batches` (`time_completed`);
 CREATE INDEX `batches_billing_project_state` ON `batches` (`billing_project`, `state`);
 
+DROP TABLE IF EXISTS `job_groups`;
+CREATE TABLE IF NOT EXISTS `job_groups` (
+  `batch_id` BIGINT NOT NULL,
+  `job_group_id` INT NOT NULL,
+  `user` VARCHAR(100) NOT NULL,
+  `attributes` TEXT,
+  `cancel_after_n_failures` INT DEFAULT NULL,
+  `state` ENUM('running', 'complete') NOT NULL,
+  `n_jobs` INT NOT NULL,
+  `time_created` BIGINT NOT NULL,
+  `time_completed` BIGINT,
+  `callback` VARCHAR(255),
+  PRIMARY KEY (`batch_id`, `job_group_id`),
+  FOREIGN KEY (`batch_id`) REFERENCES batches(`id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+CREATE INDEX `job_groups_user_state` ON `job_groups` (`user`, `state`);  # used to get cancelled job groups by user
+CREATE INDEX `job_groups_state_callback` ON `job_groups` (`batch_id`, `state`, `callback`);  # used in callback on job group completion
+CREATE INDEX `job_groups_time_created` ON `job_groups` (`batch_id`, `time_created`);  # used in list job groups and UI
+CREATE INDEX `job_groups_time_completed` ON `job_groups` (`batch_id`, `time_completed`);  # used in list job groups and UI
+CREATE INDEX `job_groups_state_cancel_after_n_failures` ON `job_groups` (`state`, `cancel_after_n_failures`);  # used in cancelling any cancel fast job groups
+
+DROP TABLE IF EXISTS `job_group_parents`;
+CREATE TABLE IF NOT EXISTS `job_group_parents` (
+  `batch_id` BIGINT NOT NULL,
+  `job_group_id` INT NOT NULL,
+  `parent_id` INT NOT NULL,
+  `level` INT NOT NULL,
+  PRIMARY KEY (`batch_id`, `job_group_id`, `parent_id`),
+  FOREIGN KEY (`batch_id`, `job_group_id`) REFERENCES job_groups (`batch_id`, `job_group_id`) ON DELETE CASCADE,
+  FOREIGN KEY (`batch_id`, `parent_id`) REFERENCES job_groups (`batch_id`, `job_group_id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+CREATE INDEX `job_group_parents_parent_id_level` ON `job_group_parents` (`batch_id`, `parent_id`, `level`);
+CREATE INDEX `job_group_parents_job_group_id_level` ON `job_group_parents` (`batch_id`, `job_group_id`, `level`);
+
 CREATE TABLE IF NOT EXISTS `batch_updates` (
```

Comment:
Do I understand correctly that, in this PR, we are setting up database structures for tracking states, cancelling, and billing but *we are not adding* structure of creation of new groups. Instead, in this PR, every batch has just one group (so we can piggy back on batch creation).

Reply:
(I made this note here because batch_updates has no job_group_id).

Reply:
Yes, you understand correctly. This is just to get job groups in the database and mirroring the correct state for the root job group and the batch.

`batch_updates` should never have a job group id as an update is a separate concept and an update can contain multiple job groups. The update is just the staged "transaction" of changes to be made to the batch rather than the job organization.